From 4beb287319b577249f24884fcf55dfaf8ba174b9 Mon Sep 17 00:00:00 2001
From: Jaroslav Kysela <jkysela@redhat.com>
Date: Wed, 3 Sep 2014 07:41:30 -0400
Subject: [alsa] hda: restore the gpio led after resume

Message-id: <140972823573-426-patchset-jkysela@redhat.com>
Patchwork-id: 91122
O-Subject: [RHEL 7.1 PATCH 425/442] ALSA: hda - restore the gpio led after resume
Bugzilla: 1112200
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>

On some HP laptops, the mute led is controlled by codec gpio.

When some machine resume from s3/s4, the codec gpio data will be
cleared to 0 by BIOS:
Before suspend:
IO[3]: enable=1, dir=1, wake=0, sticky=0, data=1, unsol=0
After resume:
IO[3]: enable=1, dir=1, wake=0, sticky=0, data=0, unsol=0

To skip the AFG node to enter D3 can't fix this problem.

A workaround is to restore the gpio data when the system resume
back from s3/s4. It is safe even on the machines without this
problem.

BugLink: https://bugs.launchpad.net/bugs/1358116
Tested-by: Franz Hsieh <franz.hsieh@canonical.com>
Cc: stable@vger.kernel.org
Signed-off-by: Hui Wang <hui.wang@canonical.com>
Signed-off-by: Takashi Iwai <tiwai@suse.de>

(picked from kernel: f475371aa65de84fa483a998ab7594531026b9d9)
rhbz#1112200
Signed-off-by: Jaroslav Kysela <jkysela@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/sound/pci/hda/patch_realtek.c b/sound/pci/hda/patch_realtek.c
index b32ce08..d71270a 100644
--- a/sound/pci/hda/patch_realtek.c
+++ b/sound/pci/hda/patch_realtek.c
@@ -3268,6 +3268,15 @@ static int alc269_resume(struct hda_codec *codec)
  snd_hda_codec_resume_cache(codec);
  alc_inv_dmic_sync(codec, true);
  hda_call_check_power_status(codec, 0x01);
+
+ /* on some machine, the BIOS will clear the codec gpio data when enter
+  * suspend, and won't restore the data after resume, so we restore it
+  * in the driver.
+  */
+ if (spec->gpio_led)
+  snd_hda_codec_write(codec, codec->afg, 0, AC_VERB_SET_GPIO_DATA,
+       spec->gpio_led);
+
  if (spec->has_alc5505_dsp)
   alc5505_dsp_resume(codec);
 
-- 
1.7.1